docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+notified about the discrepancy and the rest is burned. The amount is a
+governable parameter and can be updated at any time by the governance, with no
+governance delay.
+
+For legacy stakers, staked amount can become out-of-sync in three cases:
+
+* stake undelegated on the legacy contract,
+* stake slashed on the legacy contract,
+* stake topped-up on the legacy contract.
+
+It is expected that stake undelegation will be first performed on T staking
+contract and then on the legacy staking contract, in the same transaction.
+
+It is expected that a top-up will be first performed on the legacy staking
+contract, and then propagated to the new staking contract, in the same
+transaction. Even if it does not happen in the same transaction, this kind of
+discrepancy is not slashable given that the stake amount on the legacy contract
+is higher than the stake amount on T staking contract.
+
+In case the stake has been slashed on the legacy contract, the operator is
+required to update their information on T staking contract as soon as possible.
+In practice, with the random beacon disabled, and tBTC v1 slashing the stake
+only in case of a proven fraud that had to be committed by all operators of
+ECDSA keep, this approach is acceptable.
+
+An integral part of the staking contract should be a bot monitoring stakes and
+notifying about discrepancies. This is especially important given that the bot
+may need to voluntarily inform about discrepancies for operators that have been
+slashed to zero.
+
+==== Slashing a stake
+
+Authorized applications can slash or seize a stake. Slash operation decreases
+the stake of an operator and burns slashed tokens. Seize decreases the stake,
+burns 95% of the stake, and awards 5% to the notifier of misbehavior. 
+
+=== Limitations
+
+An absolute minimum is that existing KEEP/NU stakes should be supported in T
+network for Random Beacon, TBTC v2, and Proxy Re-Encryption. Given the fact
+supporting legacy KEEP/NU stakes forever, in every other application implemented
+for T, might not be always possible, it is expected that applications are able
+to revert authorization transactions if the delegation used in the authorization
+order does not meet the application requirements.
+
+=== Upgradeability
+
+The staking contract will be upgradeable. The exact upgradeability mechanism is
+out of the scope of this document.
+
+== Public API
+
+=== Delegating a stake
+
+==== `stake(address operator, address beneficiary, address authorizer, uint256 amount) external` 
+   
+Creates a delegation with `msg.sender` owner with the given operator,
+beneficiary, and authorizer. Transfers the given amount of T to the staking
+contract. The owner of the delegation needs to have the amount approved to
+transfer to the staking contract.
+
+==== `stakeKeep(address operator) external`
+
+Copies delegation from the legacy KEEP staking contract to T staking contract.
+No tokens are transferred. Can be called by anyone.
```

Comment:
We don't need to let anyone call this, even though we can't determine the ultimate owner of a KEEP stake easily; the operator's authoriser needs to approve the T staking contract on the KEEP staking contract (otherwise T couldn't slash), and this is exactly what the authoriser role exists for.

Thus we can limit `stakeKeep()` to "operator only".

Reply:
I think it makes sense only if this call is executed by an off-chain client. Operator keys should not leave machines clients are deployed on. This is doable but annoying - the client would need to keep an old staking contract binding and check on the start old staking contract and new staking contract. We'd also have to cover the path when the stake was "ported" but undelegated/recovered. How much iffy does it feel?

Reply:
I guess the fact that T has to be authorised is enough to prevent the minor abuse of people getting randomly staked on T with no control over the process whatsoever, so we actually don't need the access controls here

Reply:
Question regarding this method (also related to `stakeNu`):
what if operator is already in use (by staking T previously) and/or beneficiary and authorizer are different comparing between Keep contract values and T contract values? Do we want to reject tx and if so what to do with possibility stealing operator address?